### ISSUE TITLE
Allow for persistent unique block attributes

### ISSUE THREAD

**Issue by stacimc** [WordPress/gutenberg#29693]
## What problem does this address?
Blocks cannot use attributes to store information that is unique to the block instance but also persisted across page loads. Here's an illustrative example:

I'm working on the Jetpack `Pay with PayPal` block. It contains a `productId` attribute which uniquely identifies the actual product record referenced by the block. When a new block is inserted this attribute will be undefined; we hit the API to create a new product, and store the returned id in this attribute.

When the block is duplicated, I'd like to create a _new_ product record with identical fields (title/price/etc). But because the `productId` is cloned along with all the other attributes, both blocks will point to the same resource. It's possible to keep track of the block's unique `clientId` and reset the product when this value changes — but while this detects block duplication, it also changes every time the post loads, and I only want to reset the product on duplication.

## What is your proposed solution?
One option would be to add support for `unique` attributes, which do not get copied during block duplication. Another option would be to allow blocks to listen for/handle the duplication event.

**Comment by Tropicalista** [comment]
I have similar problem.

Here's my issue: https://github.com/WordPress/gutenberg/issues/29694

I'm creating a form builder. I need a way to store reference to form id in database.  When a block is made reusable it stores  the reference, but if a user convert to normal block I need a way to detach the old reference in DB and create a new one. 

I've just discovered that there's a new `__experimentalIsEditingReusableBlock` that probably can suite for my use case.

**Comment by ntsekouras** [comment]
I've been thinking about this and a possible solution I had in mind to try out involves these: 

1.  Set the unique attribute value on load only if it doesn't exist and mark this change as persistent or not depending on your needs (add undo level). Something like this: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/query/edit/index.js#L94
2. Use the `__experimentalRole` attribute to something like `id` or `unique` etc.., and then use the existing `__experimentalGetBlockAttributesNamesByRole` util to get them.
3. Handle them in the appropriate places, like in the duplicate function and where is needed.

All the above are using experimental APIs though which were introduced here: https://github.com/WordPress/gutenberg/pull/30469. I guess this could be okay though as the new functionality should also be experimental for start.

**Comment by talldan** [comment]
This could also be relevant for the widget editors, which store an internal `widgetId` as an attribute. (cc @kevin940726 @noisysocks).

**Comment by gziolo** [comment]
> This could also be relevant for the widget editors, which store an internal `widgetId` as an attribute. (cc @kevin940726 @noisysocks).

The related PR https://github.com/WordPress/gutenberg/pull/28379. We discussed the potential new APIs for block attributes with @spacedmonkey on Twitter. Related thread:

https://twitter.com/thespacedmonkey/status/1435530281595846663

> Well this is a question. I think they could be two kind of attribute. Internal and private/sensitive. Internal attributes, could used for data that is not sensitive but not design to be public. Private / sensitive, like api keys, personal information (like email) or passwords.

The conclusion in the discussion is that "internal" attribute is more something stored in the memory during the block editor's runtime and `__internalWidgetId` from the linked PR would be a good example. Here, @stacimc desribes the use case where an attribute (`productId`) should be persisted between the page edits but doesn't necessarily need to be secret. So I have a feeling that it's something between the shared definition of `internal` and `private/sensitive` 😄 Well, unless we introduce three types:
- `internal` - in memory only, not persisted in HTML, not shared when duplicating
- `private` - persisted in the HTML, not shared when duplicating
- `secret` - persisted in secure location, not shared when duplicating

**Comment by noisysocks** [comment]
> * `secret` - persisted in secure location, not shared when duplicating

Not sure I understand this. What would a "secure location" be?

**Comment by gziolo** [comment]
> > * `secret` - persisted in a secure location, not shared when duplicating
> 
> Not sure I understand this. What would a "secure location" be?

I don’t know yet, too 😅 I understood that the value couldn’t be serialized to content as HTML, so it isn’t easily exposed. I also assume it should live somewhere in the database. I was trying to document potential options and their subtle differences.

**Comment by stacimc** [comment]
I've just opened #34750 borrowing the `internal` terminology from @gziolo 😄

I think this issue may be trying to do too many things, largely because of my use of the loaded term `unique`. To solve my own original use case, it's sufficient to have a way to prevent an attribute from being copied on block duplication. `internal` feels like a superior descriptor.

Per @ntsekouras I like internal as an `__experimentalRole` for an attribute, rather than a boolean property. I'm interpreting the internal role strictly as: "attribute does not get copied on block duplication", rather than adding any additional constraints of uniqueness/etc.

I think this would go a long way to addressing many of the use cases described here, excluding the `private` or `secret` ideas. I'd like to edit the title and description of this issue slightly to replace the loaded `unique` terminology, and perhaps open a separate issue to track the desire for private/secure attributes, if that makes sense to others here. What do you think?

**Comment by gziolo** [comment]
> I've just opened #34750 borrowing the `internal` terminology from @gziolo 😄

Thank you for a quick follow-up 👍🏻 

 > I think this would go a long way to addressing many of the use cases described here, excluding the `private` or `secret` ideas. I'd like to edit the title and description of this issue slightly to replace the loaded `unique` terminology, and perhaps open a separate issue to track the desire for private/secure attributes, if that makes sense to others here. What do you think?

Feel free to edit the issue title and description according to the current direction of the PR you opened. It sounds like a good idea to open another issue to continue the discussion for other common scenarios that block authors have to deal with.

**Comment by Tropicalista** [comment]
Any update on this?

**Comment by albanyacademy** [comment]
Does copy+paste count as duplication event?

**Comment by noisysocks** [comment]
Some other use cases for an API such as this were discussed in https://github.com/WordPress/gutenberg/issues/23377.

**Comment by anver** [comment]
Any solution to this issue ?

I’m creating a custom block and one of it’s attribute is a unique id which doesn’t have any UI, so that’s not editable by the user. It’s created automatically by useEffect hook on the component startup if that attribute is not set. If I duplicate the block then the attribute is copied too which makes a duplicate id. Is there any way so that I can detect if a block is being duplicated and generate a unique id and overwrite the duplicate id with the generated one programmatically ?

**Comment by ntsekouras** [comment]
>Any solution to this issue ?

Unfortunately not yet. Your use case is exactly the same with Query Loop block and `queryId`.

**Comment by acketon** [comment]
@anver I had a similar problem for our Accordian block. We needed to generate unique ID's for use on the frontend. These were needed for aria labels and some frontend JS, etc.  I put together a work around that appears to work. It's hacky but I think it will suffice for our needs for now. It's based on some great discussion here in a few issues and on what was done in the core heading block for generating anchors. 

Basically uses useEffect() combined with searching all the blocks in the editor to check for a duplicate id stored as a data attribute. If a duplicate is found it generates a new unique ID. That seems so far to work when a block is duplicated or copy/pasted, etc. https://github.com/bu-ist/bu-blocks/issues/355. 

Example:
https://github.com/bu-ist/bu-blocks/pull/356

**Comment by anver** [comment]
> @anver I had a similar problem for our Accordian block. We needed to generate unique ID's for use on the frontend. These were needed for aria labels and some frontend JS, etc. I put together a work around that appears to work. It's hacky but I think it will suffice for our needs for now. It's based on some great discussion here in a few issues and on what was done in the core heading block for generating anchors.
> 
> Basically uses useEffect() combined with searching all the blocks in the editor to check for a duplicate id stored as a data attribute. If a duplicate is found it generates a new unique ID. That seems so far to work when a block is duplicated or copy/pasted, etc. [bu-ist/bu-blocks#355](https://github.com/bu-ist/bu-blocks/issues/355).
> 
> Example: [bu-ist/bu-blocks#356](https://github.com/bu-ist/bu-blocks/pull/356)

@acketon Thanks for the right direction, i had to change few things on that code to make it work, but i had to ditch that completely and go for a reliable solution, finally I found a solution :) Thx for the comments bro.

```ts
const getBlockCount = () => select( 'core/block-editor' ).getGlobalBlockCount();
let blockCount = getBlockCount();

	useEffect( () => {
		const unsub = subscribe( () => {
			const newCount = getBlockCount();

			if ( newCount === blockCount ) return;

			blockCount = newCount;

			const wrapper = select( 'core/block-editor' ).getBlock( clientId )!;

			const formBlocks = wrapper.innerBlocks.filter(
				( block ) => block.name === 'surveyboss/form'
			);

			const duplicates = formBlocks.filter( ( obj, index, arr ) =>
				arr.find(
					( innerObj ) =>
						innerObj.attributes.formUuid ===
							obj.attributes.formUuid &&
						innerObj.clientId !== obj.clientId
				)
			);

			if ( duplicates.length <= 1 ) return;

			dispatch( 'core/block-editor' ).updateBlockAttributes(
				duplicates[ 1 ].clientId,
				{ formUuid: generateId() }
			);

			dispatch( 'core/block-editor' ).updateBlockAttributes(
				duplicates[ 1 ].clientId,
				{
					metadata: {
						...duplicates[ 1 ].attributes.metadata,
						name: 'Untitled Form',
					},
				}
			);
		} );

		return unsub;
	}, [ clientId ] );
```

This is the part of the code i implemented and it just works fine. You got to put the effect in the parent of the duplicate components you are working with.

**Comment by Mamaduka** [comment]
It would be great to resume work on this. It appears that work was stalled due to the naming of the new attribute role and the attempt to make it configurable.

Perhaps we should follow the "local" (which is the same as `internal`, IMO) attribute example (#63076) and introduce a solution that addresses immediate problems. 

## What I'm proposing:

* Introduce a `unique` attribute role.
* Create a separate discussion regarding attribute supports for Copy, Duplicate, and Split actions.

cc @WordPress/gutenberg-core

**Comment by ramonjd** [comment]
> Introduce a unique attribute role.

Another use case (maybe) do you think this attribute could be used to tell a block that it's in "content only" mode. I'm quoting @jeryj from this [comment](https://github.com/WordPress/gutenberg/issues/58233#issuecomment-3258723872)

> Asking block authors to do const blockEditingMode = useBlockEditingMode() and then checking to display which tools is overly-complicated, messy to read, locks us into the modes idea (what if we end up adding to or changing this concept?), and leads to code duplication.

An internal prop could take away a bit of that effort.

**Comment by Mamaduka** [comment]
Sorry, @ramonjd! I don't see a connection with @jeryj's proposal. Based on PoC, it's about declarative block controls, not block attributes.

**Comment by ramonjd** [comment]
> I don't see a connection with @jeryj's proposal. Based on PoC, it's about declarative block controls, not block attributes.

Thanks for clarifying! I was barking up the wrong issue there.

**Comment by gziolo** [comment]
One note, `internal` and `content` don’t intersect with each other because the first exists only in memory on the client while the latter gets stored in the saved block. Unique is different because it can apply in both cases while enforcing the behavior so I don’t think it fits to `role`.
It’s been a long time since I interacted with the ideas around it but we should have plenty of discussion, PRs and possibly other issues to bring it to the finish line.

**Comment by Mamaduka** [comment]
I think the attribute with the `unique` role is something that gets autogenerated and isn't editable by the user. But I see your point. It seems that the "HTML Anchor" or "ID" attributes would fall in both categories. Later, primarily for block bindings.

P.S. There's no `internal` attribute role, but `local` serves the same purpose - https://github.com/WordPress/gutenberg/pull/63076.

**Comment by gziolo** [comment]
> P.S. There's no internal attribute role, but local serves the same purpose - https://github.com/WordPress/gutenberg/pull/63076.

Right, I said it’s been a while. I forgot the exact naming used 😀

**Comment by adamsilverstein** [comment]
This issue came up in work on block commenting where we need  to avoid two blocks linking to the same (block scoped) comment thread. This happens when splitting a block as well as when copying in the same post I assume.

**Comment by mtias** [comment]
@adamsilverstein would it make sense in some cases to actually support that? I'm thinking of an article where an editor comments on something that affects multiple images in the story and there may be a single unique thread attached or referenced by each of the image blocks. May be a bit niche, but it sounds like it could have valid and useful scenarios. It's somewhat common on gdocs to see a "same feedback here as in the other comment" :D

**Comment by mcsf** [comment]
> would it make sense in some cases to actually support that? I'm thinking of an article where an editor comments on something that affects multiple images in the story and there may be a single unique thread attached or referenced by each of the image blocks. May be a bit niche, but it sounds like it could have valid and useful scenarios. It's somewhat common on gdocs to see a "same feedback here as in the other comment" :D

This is valid, but we'd need to fix a few things if we want to purse it. Opening the console reveals React warnings about duplicate keys, so we'd probably rekey notes as `${blockId}-${noteId}`. The focus rules would need adjustment, etc. But, more importantly, I wonder about the UX: will users be able to unmap blocks, and how will that be presented without putting a large cognitive burden on users?
